vcd/resource_vcd_nsxv_firewall_rule_test.go

```diff
@@ -0,0 +1,879 @@
+func checkfirewallRuleOrder(firstRule, secondRule string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*VCDClient)
+		edgeGateway, err := conn.GetEdgeGateway(testConfig.VCD.Org, testConfig.VCD.Vdc, testConfig.Networking.EdgeGateway)
+		if err != nil {
+			return fmt.Errorf(errorUnableToFindEdgeGateway, err)
+		}
+
+		rule1, ok := s.RootModule().Resources[firstRule]
+		if !ok {
+			return fmt.Errorf("not found resource: %s", firstRule)
+		}
+
+		rule2, ok := s.RootModule().Resources[secondRule]
+		if !ok {
+			return fmt.Errorf("not found resource: %s", secondRule)
+		}
+
+		rule1Id := rule1.Primary.ID
+		rule2Id := rule2.Primary.ID
+
+		allFirewallRules, err := edgeGateway.GetAllNsxvFirewallRules()
+		if err != nil {
+			return fmt.Errorf("could not get all firewall rules: %s", err)
+		}
+
+		// This loop checks that at first rule1Id is found and then rule2Id is found in the ordered
+		// list of firewall rules
+		var foundSecondRule, foundFirstRule bool
+		for _, rule := range allFirewallRules {
+			if rule.ID == rule1Id {
+				foundFirstRule = true
+			}
+			if foundFirstRule && rule.ID == rule2Id {
+				foundSecondRule = true
+			}
+		}
+
+		if !foundSecondRule || !foundFirstRule {
+			return fmt.Errorf("incorrect rule order. %s is above %s. Should be reverse.", firstRule, secondRule)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckVcdFirewallRuleDestroy(resource string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resource]
+		if !ok {
+			return fmt.Errorf("not found resource: %s", resource)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("no ID is set for %s resource", resource)
+		}
+
+		conn := testAccProvider.Meta().(*VCDClient)
+
+		edgeGateway, err := conn.GetEdgeGateway(testConfig.VCD.Org, testConfig.VCD.Vdc, testConfig.Networking.EdgeGateway)
+		if err != nil {
+			return fmt.Errorf(errorUnableToFindEdgeGateway, err)
+		}
+
+		rule, err := edgeGateway.GetNsxvFirewallRuleById(rs.Primary.ID)
+
+		if !govcd.IsNotFound(err) || rule != nil {
+			return fmt.Errorf("firewall rule (ID: %s) was not deleted: %s", rs.Primary.ID, err)
+		}
+		return nil
+	}
+}
+
+const testAccVcdEdgeFirewallRule0 = `
+resource "vcd_nsxv_firewall_rule" "rule0" {
+	org          = "{{.Org}}"
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+	name = "test-rule"
+	rule_tag = "30000"
+	action = "deny"
+
+	source {
+		ip_addresses = ["any"]
+	}
+  
+	destination {
+		ip_addresses = ["192.168.1.110"]
+	}
+  
+	service {
+		protocol = "any"
+	}
+}
+
+resource "vcd_nsxv_firewall_rule" "rule0-2" {
+	org          = "{{.Org}}"
+	vdc          = "{{.Vdc}}"
+	edge_gateway = "{{.EdgeGateway}}"
+	name = "rule 123123"
+	action = "deny"
+
+	source {
+		ip_addresses = ["4.4.4.4"]
+	}
+  
+	destination {
+		ip_addresses = ["5.5.5.5"]
+	}
+  
+	service {
+		protocol = "any"
+	}
+	depends_on = ["vcd_nsxv_firewall_rule.rule0"]
```

Comment:
Why we need this here?

Reply:
This is one way of ordering firewall rules (besides using the `above_rule_id`). Line 141 has a check that `rule0` is created first in the order before `rule0-2`

`checkfirewallRuleOrder("vcd_nsxv_firewall_rule.rule0", "vcd_nsxv_firewall_rule.rule0-2"),`

Reply:
ok.